include/deal.II/lac/solver_fire.h

```diff
@@ -0,0 +1,392 @@
+      u.add (timestep, velocities);                // U += dt     * V
+      inverse_masses.vmult(gradients, gradients);  // G  = M^{-1} * G
+      velocities.add (-timestep, gradients);       // V -= dt     * G
+
+      // Compute gradients for the new u.
+      compute(gradients, u);
+
+      const real_type gradient_norm_squared = gradients * gradients;
+      conv = this->iteration_status(iter, gradient_norm_squared, u);
+      if (conv != SolverControl::iterate)
+        break;
+
+      // v_dot_g = V * G
+      const real_type v_dot_g = velocities * gradients;
+
+      if (v_dot_g < 0.)
+        {
+          const real_type velocities_norm_squared =
+            velocities * velocities;
+
+          // Check if we divide by zero in DEBUG mode.
+          Assert (gradient_norm_squared > 0., ExcInternalError());
+
+          // beta = - alpha |V|/|G|
+          const real_type beta = -alpha *
+                                 std::sqrt (velocities_norm_squared
+                                            /
+                                            gradient_norm_squared);
+
+          // V = (1-alpha) V + beta G.
+          velocities.sadd (1. - alpha, beta, gradients);
+
+          if (iter - previous_iter_with_positive_v_dot_g > DELAYSTEP)
+            {
+              // Increase timestep and decrease alpha.
+              timestep = std::min (timestep*TIMESTEP_GROW, maximum_timestep);
+              alpha *= ALPHA_SHRINK;
+            }
+        }
+      else
+        {
+          // Decrease timestep, reset alpha and set V = 0.
+          previous_iter_with_positive_v_dot_g = iter;
+          timestep *= TIMESTEP_SHRINK;
+          alpha = ALPHA_0;
+          velocities = 0.;
+        }
+
+      real_type vmax = velocities.linfty_norm();
+
+      // Change timestep if any dof would move more than maximum_linfty_norm.
+      if (vmax > 0.)
+        {
+          const double minimal_timestep = additional_data.maximum_linfty_norm
+                                          /
+                                          vmax;
+          if (minimal_timestep < timestep)
+            timestep = minimal_timestep;
+        }
+
+      print_vectors(iter, u, velocities, gradients);
+
+    } // While we need to iterate.
+
+  deallog.pop();
+
+  // In the case of failure: throw exception.
+  if (conv != SolverControl::success)
+    AssertThrow (false,
+                 SolverControl::NoConvergence (iter, gradients * gradients));
+
+}
+
+
+
+template <typename VectorType>
+template<typename MatrixType, typename PreconditionerType>
+void SolverFIRE<VectorType>::solve (const MatrixType         &A,
+                                    VectorType               &x,
+                                    const VectorType         &b,
+                                    const PreconditionerType &precondition)
+{
+
+  std::function<double(VectorType &,  const VectorType &)> compute_func =
+    [&]               (decltype(x) &G, decltype(b)       &x) -> double
```

Comment:
`decltype(x)` and `decltype(b)` are actually both `const VectorType` here -- just use that type.